### ISSUE TITLE
Allow to return null in BiFunction argument of Map.compute

### ISSUE THREAD

**Issue by olegshtch** [typetools/checker-framework#2460]
Next code doesn't compile:

```java
import java.util.Map;
import java.util.Set;
import java.util.HashMap;

class M {
	private final Map<String, Set<String>> storage = new HashMap<>();
	
	void doSomething() {
		String key = "someString";
		storage.compute(key, (k, v) -> {
			v.remove(k);
			if (v.isEmpty()) {
				return null;
			} else {
				return v;
			}
		});
	}
}
```

While javadoc https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#compute-K-java.util.function.BiFunction- says BiFunction could return `null` to remove entry from map.

jdk1.8.0_202, checkerfamework 2.5.7

**Comment by wmdietl** [comment]
The following modification of your example passes the Nullness Checker:

```java
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import org.checkerframework.checker.nullness.qual.Nullable;

class M {
	private final Map<String, @Nullable Set<String>> storage = new HashMap<>();
	
	void doSomething() {
		String key = "someString";
		storage.compute(key, (k, v) -> {
			if (v == null) {
                        	return null;
			}
			v.remove(k);
			if (v.isEmpty()) {
				return null;
			} else {
				return v;
			}
		});
	}
}
```

Note that now the declaration of `storage` makes explicit that the Set can be null.
This then also forces you to account for a null set before trying to remove the value.

Is this what you want? If so, I think all the APIs are correctly annotated to allow this use.

**Comment by olegshtch** [comment]
I don't want to store `null`s in a `Map`. I want to remove entry if I return `null` in `BiFunction`.

**Comment by olegshtch** [comment]
I've added astub file which fixes it to my project:
```java
package java.util;

import org.checkerframework.checker.nullness.qual.Nullable;

interface Map<K, V> {

    V computeIfAbsent(K arg0, Function<? super K,@Nullable ? extends @Nullable V> arg1);
    V computeIfPresent(K arg0, BiFunction<? super K,? super V,@Nullable ? extends @Nullable V> arg1);
    V compute(K arg0, BiFunction<? super K,? super V,@Nullable ? extends @Nullable V> arg1);
    V merge(K arg0, V arg1, BiFunction<? super V,? super V,@Nullable ? extends @Nullable V> arg2);
}

**Comment by wmdietl** [comment]
Ah, sorry, I misunderstood.
Using `@Nullable ? extends @Nullable V` seems overly restrictive, as it now requires a method that returns `null`. Whouldn't simply `? extends @Nullable V` allow your use case and allow a function that has a `@NonNull Value` return type?

**Comment by olegshtch** [comment]
I've just tested it. `? extends @Nullable V` works as well.

**Comment by wmdietl** [comment]
Thanks for confirming!
Would you be willing to add your stub annotations to https://github.com/typetools/checker-framework/blob/master/checker/jdk/nullness/src/java/util/Map.java in a pull request?

**Comment by olegshtch** [comment]
@wmdietl Done!

**Comment by olegshtch** [comment]
Looks like this issue return back.
